fundraiser/src/main.sw

```diff
@@ -0,0 +1,448 @@
+        // checks
+        require(!campaign_info.cancelled, CampaignError::CampaignHasBeenCancelled);
+
+        // Mark the campaign as claimed and overwrite the previous state with the updated version
+        campaign_info.claimed = true;
+        storage.campaign_info.insert(id, campaign_info);
+
+        // Transfer the total pledged to this campaign to the beneficiary
+        transfer(campaign_info.total_pledge, campaign_info.asset, campaign_info.beneficiary);
+
+        // We have updated the state of a campaign therefore we must log it
+        log(ClaimedEvent {
+            id
+        });
+    }
+
+    /// Allows a user to pledge any amount of the campaign asset towards the campaign goal
+    ///
+    /// In order to reach the campaign's target amount users must pledge some amount of asset towards
+    /// that campaign.
+    /// This information is recorded for the campaign and for the user so that they can unpledge.
+    ///
+    /// # Arguments
+    ///
+    /// * `id` - Unique campaign identifier which is a number from the storage.total_campaigns range
+    ///
+    /// # Reverts
+    ///
+    /// * When the `id` is either 0 or greater than the total number of campaigns created
+    /// * When the user attempts to pledge when the deadline has been reached
+    /// * When the user pledges a different asset to the one specified in the campaign
+    /// * When the user pledges after the campaign has been cancelled
+    /// * When an AuthError is generated
+    #[storage(read, write)]fn pledge(id: u64) {
+        // User cannot interact with a non-existent campaign
+        validate_id(id, storage.total_campaigns);
+
+        // Retrieve the campaign in order to check its data / update it
+        let mut campaign_info = storage.campaign_info.get(id);
+
+        // The users should only have the ability to pledge to campaigns that have not reached their
+        // deadline (ended naturally - not been cancelled)
+        require(height() < campaign_info.deadline, CampaignError::CampaignEnded);
+
+        // The campaign specifies an asset that it accepts therefore the user must pledge the correct
+        // asset in order to update the state of the campaign
+        require(campaign_info.asset == msg_asset_id(), UserError::IncorrectAssetSent);
+
+        // A user cannot pledge zero since it does not make sense to do so
+        require(0 < msg_amount(), UserError::AmountCannotBeZero);
+
+        // The user should not be able to continue to pledge if the campaign has been cancelled
+        // Given the logic below it's unnecessary but it makes sense to stop them
+        require(!campaign_info.cancelled, CampaignError::CampaignHasBeenCancelled);
+
+        // Use the user's pledges as an ID / way to index this new pledge
+        let user = sender_identity();
+        let pledge_count = storage.pledge_count.get(user);
+
+        // First time pledging to any campaign
+        if pledge_count == 0 {
```

Comment:
Is checking for 0 necessary? if `pledge_count` is 0 then `storage.pledge_count.insert(user, pledge_count + 1);` from below would accomplish the same thing as `storage.pledge_count.insert(user, 1);`

Reply:
Yeah, you're correct. My reasoning was that at first they have not pledged otherwise there is a pledge but it might be to a new campaign or an existing campaign so I would have an initial check and then another branch between those latter two. The logic can be updated (including the comments) to use the pledge index instead.